### PR TITLE
translate addresses out of VMA space on linux

### PR DESCRIPTION
**Pull request by theyy29**
This may be useful; This _may_ break non-PIE binaries.

### CODE REVIEW

**Comment by wc-duck** [comment]
I'll get to this eventually... Just haven't had the time yet!

File: src/callstack.cpp [hunk 1]
```diff
@@ -60,9 +60,98 @@ static const char* alloc_string( callstack_string_buffer_t* buf, const char* str
...
+			char *ret = fgets(buf, 4096, maps);
+			if(!ret)
+				break;
+
+			mmap_entry_t *next = (mmap_entry_t*) malloc(sizeof(mmap_entry_t));
+			next->next = NULL;
+
+			// from man `proc`:
+			// address perms offset dev inode pathname
+			sscanf(buf, "%lx-%lx %ms %lx %ms %ms %ms\n",
```
Comment (wc-duck) [inline src/callstack.cpp:102]: there is really no need for this lib to dynamically alloc all the strings as you don't use them, so I would just change the %ms to %*s and remove the string-members from mmap_entry_t.

File: src/callstack.cpp [hunk 2]
```diff
@@ -82,9 +171,21 @@ static const char* alloc_string( callstack_string_buffer_t* buf, const char* str
 	#   error "Unhandled platform"
 	#endif
 
+	#if defined(__linux)
+		mmap_entry_t *list = parse_mmaps("/proc/self/maps");
```
Comment (wc-duck) [inline src/callstack.cpp:175]: It is a small thing, but maybe it is worth just allocating x mmap_entry_t on the stack. I'm not really sure how many entries might show up in /proc/self/maps? But looking at my machine it is not "that much".
That being said, I'll accept this as well and maybe fix it up later if I need to.

**Review by wc-duck — COMMENTED** [review]
I just had the comments that you find in the review... I'll fix them myself you you don't want to. And a big thanks for taking the time and fixing this!

File: src/callstack.cpp (hunk 1 above)
Comment (theyy29) [reply src/callstack.cpp]: Fixed. I copied this from something else I wrote that did use (some of) them

File: src/callstack.cpp (hunk 2 above)
Comment (theyy29) [reply src/callstack.cpp:161]: Well, something on my system (firefox) has 2000+ lines... More interesting to me would be python, which looks like it starts pretty small (70), but as you import more things, it grows.

Alternatively, if we only care about symbols from the main program (and not any libraries), which is an okay assumption, (we only run addr2line for the main program), then we only need to read the first "r-xp" line from /proc/self/maps.

Comment (wc-duck) [reply src/callstack.cpp:161]: I think this is fine for now. If it turns out to be a problem that can be fixed later.

**Pull request merged**